ui: Use BaseURL to reduce relative imports

mrtracy opened the pull request:
This commit uses typescript's `baseUrl` setting, in combination with
webpack's `resolver` settings, in order to eliminate nearly all
cross-directory relative imports (i.e. uses of "../").

In this configuration, typescript and webpack look for non-relative
imports in two locations: the '/src' directory and the '/node_modules'
directory, in that order.  This allows all cross-folder imports within
the cockroach code to be expressed with non-relative notation, which
will make upcoming code refactoring much easier.

Relative imports are still used to reference files in the same folder
(i.e. using "./" with a single dot).

There are a few uses of "../" imports remaining in the project, but they
are not importing code - they are importing non-code cockroachDB assets
which are located outside of the "/src" directory. Those asset
directories could be added to the resolution order in webpack, but since
they are so few in number I would like to continue with upcoming code
refactors to see how the directory structure is changed.

Commit:
ui: Use BaseURL to reduce relative imports

This commit uses typescript's `baseUrl` setting, in combination with
webpack's `resolver` settings, in order to eliminate cross-directory
relative imports (i.e. uses of "../").

In this configuration, typescript and webpack look for non-relative
imports in two locations: the top level '/ui' directory and the
'/node_modules' sub-directory, in that order.  This allows all
cross-folder imports within the cockroach code to be expressed with
non-relative notation, which will make upcoming code refactoring much
easier.

Relative imports are still used to reference files in the same folder
(i.e. using "./" with a single dot).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16074)
<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 54 of 54 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ui/tsconfig.json, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmQFhwThpBtNtcjt2j:-KkmQFhwThpBtNtcjt2k:betzzga) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/tsconfig.json#L21)):*
> ```JSON
>     "baseUrl": ".",
>     "paths": {
>       "*" : [
> ```

nit: remove the space before the colon

---

*[pkg/ui/tsconfig.json, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmOtBHvh38pd0BtqQe:-KkmOtBHvh38pd0BtqQf:b-3crsgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/tsconfig.json#L22)):*
> ```JSON
>     "paths": {
>       "*" : [
>         "node_modules/*",
> ```

How about:

```
"*": [
  "*",
  "node_modules/*"
]
```

if that works.

likewise for the webpack config

---

*[pkg/ui/webpack.config.js, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmP3q0Ne9LKa8W0jXr:-KkmP3q0Ne9LKa8W0jXs:b-ue24hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/webpack.config.js#L5)):*
> ```JavaScript
> const path = require('path');
> const rimraf = require('rimraf');
> ```

nit: random newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-:-KkmQ_7gWnXt28Z5h_Xd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 2 of 56 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ui/tsconfig.json, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmQFhwThpBtNtcjt2j:-Kkq_YkEsoWNVJ8grAT2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/tsconfig.json#L21)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: remove the space before the colon
</blockquote></details>

Done.

---

*[pkg/ui/tsconfig.json, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmOtBHvh38pd0BtqQe:-Kkq_Z8tZvG30hCWdzFY:b-2us4e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/tsconfig.json#L22)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

How about:

```
"*": [
  "*",
  "node_modules/*"
]
```

if that works.

likewise for the webpack config
</blockquote></details>

Done, also rewrote all project imports as "src/" accordingly.  As a result of this, I was also able to eliminate the remaining "../" imports that were mentioned in the original commit message.  I have updated the commit message accordingly.

---

*[pkg/ui/webpack.config.js, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-KkmP3q0Ne9LKa8W0jXr:-Kkq_rbeoY3UaRKYEizK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/632cc1af28bc9cff7623bffbd25e704c35f1613b/pkg/ui/webpack.config.js#L5)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: random newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074)*
<!-- Sent from Reviewable.io -->

tamird commented:
:tada: 

---

Reviewed 54 of 54 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16074#-:-KkqeNox6LGF3V2tBcwJ:bjwa7hh)*
<!-- Sent from Reviewable.io -->
